internal/terraform/node_resource_partial_plan.go

```diff
@@ -0,0 +1,318 @@
+		return &Graph{}, nil, diags
+	}
+
+	// Our graph transformers require access to the full state, so we'll
+	// temporarily lock it while we work on this.
+	state := ctx.State().Lock()
+	defer ctx.State().Unlock()
+
+	maybeOrphans := addrs.MakeSet[addrs.AbsResourceInstance]()
+
+	steps := []GraphTransformer{
+
+		DynamicTransformer(func(graph *Graph) error {
+			// We'll add a node for all the known instance keys.
+			for _, key := range knownInstKeys {
+				graph.Add(n.concreteResource(imports, n.skipPlanChanges)(NewNodeAbstractResourceInstance(addr.Instance(key))))
+			}
+			return nil
+		}),
+
+		DynamicTransformer(func(graph *Graph) error {
+			// We'll add a node if there are unknown instance keys.
+			if haveUnknownKeys {
+				graph.Add(&nodePlannablePartialExpandedResource{
+					addr:             addr.Module.UnexpandedResource(addr.Resource),
+					config:           n.Config,
+					resolvedProvider: n.ResolvedProvider,
+					skipPlanChanges:  n.skipPlanChanges,
+				})
+			}
+			return nil
+		}),
+
+		DynamicTransformer(func(graph *Graph) error {
+			// We'll add nodes for any orphaned resources.
+			rs := state.Resource(addr)
+		Instances:
+			for key, inst := range rs.Instances {
+				if inst.Current == nil {
+					continue
+				}
+
+				for _, knownKey := range knownInstKeys {
+					if knownKey == key {
+						// Then we have a known instance, so we can skip this
+						// one - it's definitely not an orphan.
+						continue Instances
+					}
+				}
+
+				if haveUnknownKeys {
+					// Then this is a "maybe orphan" instance. It isn't mapped
+					// to a known instance but we have unknown keys so we don't
+					// know for sure that it's been deleted.
+					maybeOrphans.Add(addr.Instance(key))
+					graph.Add(n.concreteResource(addrs.MakeMap[addrs.AbsResourceInstance, string](), true)(NewNodeAbstractResourceInstance(addr.Instance(key))))
+					continue
+				}
+
+				// If none of the above, then this is definitely an orphan.
+				graph.Add(n.concreteResourceOrphan(NewNodeAbstractResourceInstance(addr.Instance(key))))
+			}
+
+			return nil
+		}),
+
+		// Attach the state
+		&AttachStateTransformer{State: state},
+
+		// Targeting
+		&TargetsTransformer{Targets: n.Targets},
+
+		// Connect references so ordering is correct
+		&ReferenceTransformer{},
+
+		// Make sure there is a single root
+		&RootTransformer{},
+	}
+
+	b := &BasicGraphBuilder{
+		Steps: steps,
+		Name:  "nodeExpandPlannableResource",
+	}
+	graph, graphDiags := b.Build(addr.Module)
+	diags = diags.Append(graphDiags)
+	return graph, maybeOrphans, diags
+}
+
+// transformDynamic is a helper struct that wraps a single function, allowing
+// us to transform a graph dynamically.
+type transformDynamic struct {
+	Transformer func(*Graph) error
+}
+
+// DynamicTransformer returns a GraphTransformer that will apply the given
+// function to the graph during the dynamic expansion phase.
+func DynamicTransformer(f func(*Graph) error) GraphTransformer {
```

Comment:
Yeah, I don't think this will be long lived. When we come back and merge these two paths (which I think answers your other question with a "yes, we will eventually merge these once the alpha periods for stacks / deferred actions is finished), we'll integrate the logic in these dynamic transformers into the "proper" transformers that already exist for the alternate code paths. I just wanted to keep things separate for now, and didn't want to go and create a load of new but temporary files, instead preferring to keep all the "to-be-merged" stuff in one place.